internal/expr/bindinputs.go

```diff
@@ -45,47 +45,74 @@ func (tbe *TypeBoundExpr) BindInputs(args ...any) (pq *PrimedQuery, err error) {
 	for _, te := range *tbe {
 		switch te := te.(type) {
 		case *typedInputExpr:
-			vals, omit, err := te.input.LocateParams(typeToValue)
+			params, err := te.input.LocateParams(typeToValue)
 			if err != nil {
 				return nil, err
 			}
-			if omit {
+			if params.Omit {
 				return nil, omitEmptyInputError(te.input.Desc())
 			}
+			if params.Bulk {
+				return nil, fmt.Errorf("cannot use bulk inputs outside an insert statement")
+			}
 
-			argTypeUsed[te.input.ArgType()] = true
-			sqlBuilder.writeInput(inputCount, len(vals))
-			for _, val := range vals {
-				namedInput := sql.Named("sqlair_"+strconv.Itoa(inputCount), val.Interface())
-				params = append(params, namedInput)
+			argTypeUsed[params.ArgTypeUsed] = true
+			sqlBuilder.writeInput(inputCount, len(params.Vals))
+			for _, val := range params.Vals {
+				namedInput := sql.Named("sqlair_"+strconv.Itoa(inputCount), val)
+				namedInputs = append(namedInputs, namedInput)
 				inputCount++
 			}
 		case *typedInsertExpr:
-			var columns []string
+			var boundColumns []*boundColumn
+			var columnNames []string
+			bulk := false
+			numRows := 1
+			// firstBulkColumn stores the type name of the first column used in
+			// a bulk insert. This is used for error messages.
+			var firstBulkColumn string
 			for _, ic := range te.insertColumns {
-				vals, omit, err := ic.input.LocateParams(typeToValue)
+				bc, err := ic.bindInputs(typeToValue)
 				if err != nil {
 					return nil, err
 				}
-				if len(vals) > 1 {
-					// Only slices return multiple values and they are not
-					// allowed in insert expressions.
-					return nil, fmt.Errorf("internal error: types in insert expressions cannot return multiple values")
+				bc.markArgUsed(argTypeUsed)
+
+				if bc.bulk {
```

Comment:
These additions take the length of this method over 100 lines. We should seek ways to make it more terse.

Reply:
I've refactored this method into a new `queryBuilder`. Then used the visitor pattern with the `typedExpr` to call the add methods on it.